### PR TITLE
Let file sources choose a path for uploaded files

### PR DESCRIPTION
**Pull request by domgz**
The method `write_from()` of `SingleFileSource` and `BaseFilesSource` reads a local file from `native_path` and saves it to `target_path` on a file source.

This PR enables the service backing the file source to choose which will be the path of the saved file, meaning that `target_path` and the actual path where the file can be recovered later do not have to match. The latter is the return value of `write_from()`.

Therefore, all usages of `write_from()` have also been refactored to consider the paths chosen by the file source's backing service. In addition, when exporting a history, the URI that the service backing the file source assigns to it will be saved to the history export result metadata object (`StoreExportAssociation.export_metadata`).

The changes from this PR are meant to address a fundamental design mismatch between Galaxy and eLabFTW that impedes the smooth integration of the latter. Galaxy file sources are designed to save files to a `target_uri` which is set _before_ the file is saved, partly by the user, who chooses the basename in the URI. On the other hand, files can be retrieved from eLabFTW via an `entity_type`, `entity_id` and `upload_id`, and the latter two are set by eLabFTW. There is no good mapping between `target_uri` and `entity_type`, `entity_id` and `upload_id` that can work without breaking the assumption that "saving a file on path `x` guarantees that it can be retrieved later using `x`", which is precisely what this PR does. The issue galaxyproject/galaxy#18665 includes a more thorough explanation of this design mismatch.

This is the first PR of a series of PRs that integrate eLabFTW with Galaxy via a file source (together they address issue galaxyproject/galaxy#18665):
- galaxyproject/galaxy#19154
- galaxyproject/galaxy#19256
- galaxyproject/galaxy#19319

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jmchilton** [comment]
I've read through the attached issue and I am really nervous about this approach. It seems like there is a mismatch because the abstractions are pretty different. It seems like a augmenting the model import/export process would make more sense than this lower level abstraction. Do have the details of the plugin worked through that prompted this change? David's comment on that issue seems to vaguely point in this direction as well.

**Comment by kysrpex** [comment]
> I've read through the attached issue and I am really nervous about this approach. It seems like there is a mismatch because the abstractions are pretty different. It seems like a augmenting the model import/export process would make more sense than this lower level abstraction. Do have the details of the plugin worked through that prompted this change? David's comment on that issue seems to vaguely point in this direction as well.

Sorry for not answering earlier, I was out of office for most of the week.

Indeed there is a very large mismatch between the abstractions from Galaxy and eLabFTW. Galaxy works with a unique URI that is fixed _before_ uploading to the file source, whose basename is chosen by the user; eLabFTW works with a pair of autoincrementing identifiers that are fixed _after_ creating the experiment or resource or uploading the file and a string enum with two choices (i.e. `experiment` or `item`).

About what @davelopez [commented on the issue](https://github.com/galaxyproject/galaxy/issues/18665#issuecomment-2275653043)

> I need to study the case a bit, but as a first impression, this case clearly will need a new special UI entry here
> screenshot
> This UI will have to create the needed entities before the "export" similar to what the RDM file sources are doing. Then, once you have a proper URI that identifies the target entity (something like: elabftw://{elab_url}/entity_type/entity_id/upload_id) perform the upload in the backend. I don't know if that is possible, I haven't checked the eLabFTW API but that could be a potential solution.

You can create entities via eLabFTW, and you can have as many files attached to them as you wish, therefore you do not necessarily need to create an entity to export a file, in fact most of the time you will be interested in exporting files to an existing entity, because they have a lot of metadata that should be filled through the eLabFTW interface.

> Do have the details of the plugin worked through that prompted this change?

I am not sure if you are referring to the design of the plugin or the code of the plugin. The latter is almost ready [here](https://github.com/kysrpex/galaxyproject-galaxy/commits/eLabFTW/), although I do not think it makes sense to have a look yet (let's tackle problems step by step, there is another large problem/mismatch to discuss later). What prompts the change when working the plugin through is the moment one has to **choose a valid mapping between eLabFTW entity type, entity id and upload id and Galaxy URI** that respects their properties:

- Galaxy URI
  - Pathname fixed by Galaxy, basename chosen by the user, both _before_ exporting the file (the user-set basename is not a problem, it can just be appended to the URI and ignored)
  - File can be retrieved later on the same URI
- eLabFTW (entity type, entity id and upload id)
  - Autoincrementing entity ids fixed by eLabFTW _after_ creating the entity, next id unpredictable (can be known only by the admin, problems with concurrently creating entities)
  - Autoincrementing upload ids fixed by eLabFTW _after_ uploading the file, next id unpredictable (problems with concurrent uploads)

In other words, the simple part of the problem is "Construct an injective function from the set of nonnegative integer 2-tuples to the set of strings complying with the definition of a path according to RFC 3986". The impossible part of the problem is that Galaxy needs to apply the inverse function without having anything to apply it to!

Anyway, @davelopez wants to have closer look at this together with me to see if the change from this PR can be avoided. If a better solution can be found then fine, otherwise augmenting the model import/export process would make sense, and I do not mind putting the effort. Let's first have this closer look and then we may continue the discussion.

File: lib/galaxy/managers/model_stores.py [hunk 1]
```diff
@@ -214,21 +214,29 @@ def write_history_to(self, request: WriteHistoryTo):
...
+        finally:
+            export_metadata = self.set_history_export_request_metadata(request)
             self.set_history_export_result_metadata(
-                request.export_association_id, export_metadata, success=False, error=str(e)
+                request.export_association_id,
+                export_metadata,
+                success=not bool(exception_exporting_history),
+                error=str(exception_exporting_history) if exception_exporting_history else None,
             )
-            raise
```
Comment (domgz) [inline lib/galaxy/managers/model_stores.py]: Before applying the set of changes from 53012bf, the method `ModelStoreManager.set_history_export_request_metadata()` instantiates a `ExportObjectMetadata` Pydantic model and dumps it to the database in the form of JSON as the field `StoreExportAssociation.export_metadata`. After the export is complete, the method `set_history_export_result_metadata()` takes the same instance of `ExportObjectMetadata`, instantiates a `ExportObjectResultMetadata` Pydantic model, sets it as the `result_data` of the `ExportObjectMetadata` instance, and then saves the `ExportObjectMetadata` Pydantic model in the form of JSON to the database again.

After applying the set of changes, the call to `ModelStoreManager.set_history_export_request_metadata()` is delayed until the file has already been saved to the file source, as the actual URI that the file source assigns to the file is otherwise unknown.

The URI assigned by the file source overwrites the original target URI in the request. This involves a slight deviation from the previous behavior: if for example, power gets cut at the right time, `StoreExportAssociation.export_metadata` may not exist despite the history having been already saved to the file source, because database writes happen within the `finally:` block.

Moreover, overwriting the original target URI from the request is formally wrong, because the actual URI assigned by the file source should be part of the export result metadata, as it becomes known when the export completes. However, that implies modifying the other parts of the codebase that reference the URI from the request.

Despite the slight deviation in behavior and the formal incorrectness, rather than jumping straight into fixing these issues, I think it makes sense to leave the chance for discussion open, as doing things this way may still be an interesting tradeoff. Let me know what you think.

**Comment by jmchilton** [comment]
I appreciate you taking the time to at least simplify the implementation classes with that change being applied in the base class. I'm still very uneasy but I will live - especially if @davelopez is okay with this approach.

**Comment by davelopez** [comment]
Thanks for the trust @jmchilton :smile: :pray:  I know it is somewhat of an API change and we are streching a bit the original implementation, but I think it is safe to assume the normal case would be returning the same `target_path` from the `write_from` method in 99.9% of the cases which should be harmless. If it doesn't, the plugin should be responsible for handling it as in this particular case that motivated the change.

File: lib/galaxy/managers/model_stores.py (hunk 1 above)
Comment (domgz) [reply lib/galaxy/managers/model_stores.py]: I was thinking...

> After applying the set of changes, the call to ModelStoreManager.set_history_export_request_metadata() is delayed until the file has already been saved to the file source, as the actual URI that the file source assigns to the file is otherwise unknown.

If we merge the PR as it is, then we'd never see an export that is in progress in the list from the UI. If the file is large, that alone would justify attempting to fix

> overwriting the original target URI from the request is formally wrong, because the actual URI assigned by the file source should be part of the export result metadata, as it becomes known when the export completes

right? I guess it makes sense to make an attempt.

Comment (domgz) [reply lib/galaxy/managers/model_stores.py]: 7a9a52a scratches this itch

File: lib/galaxy/schema/schema.py [hunk 2]
```diff
@@ -1833,8 +1833,35 @@ class ExportObjectRequestMetadata(Model):
...
+
+    @model_validator(mode="after")
+    @classmethod
+    def validate_uri(cls, model):
+        """
+        Ensure unsuccessful exports do not have a URI.
+        """
+        # short therm exports need not have a URI
+        # if model.success and not model.uri:
+        #     raise ValueError("successful exports must have a URI")
```
Comment (davelopez) [inline lib/galaxy/schema/schema.py]: You can remove this commented out code :)
```suggestion
```

**Review by davelopez — APPROVED** [review]
Thank you @kysrpex for the improvements and refactorings!
This looks and works nicely!

**Comment by kysrpex** [comment]
> Thank you @kysrpex for the improvements and refactorings! This looks and works nicely!

Dankeschön! Remember I do not have permission to merge this, so it has to be one of you.
